leshan-core/src/main/java/org/eclipse/leshan/core/request/argument/Arguments.java

```diff
@@ -0,0 +1,314 @@
+        private final String digitPart;
+        private final String valueDecorated;
+        private final String content;
+
+        public ArgumentParser(String argument, String content) {
+            this.content = content;
+            String[] keyValue = argument.split("=", 2);
+            digitPart = keyValue[0];
+            valueDecorated = keyValue.length == 1 ? null : keyValue[1];
+        }
+
+        private Argument parse() throws InvalidArgumentException {
+            if (!isValid()) {
+                throw new InvalidArgumentException("Unable to parse Arguments [%s]", content);
+            }
+            int digit;
+            try {
+                digit = Integer.parseInt(digitPart);
+            } catch (NumberFormatException e) {
+                throw new InvalidArgumentException(e, "Unable to parse Arguments [%s] with digit [%s]", content, digitPart);
+            }
+            String value = null;
+            if (valueDecorated != null) {
+                value = StringUtils.removeEnd(StringUtils.removeStart(valueDecorated, "'"), "'");
+            }
+
+            return new Argument(digit, value);
+        }
+
+        private boolean isValid() {
+            if (digitPart.length() != 1) {
+                return false;
+            }
+
+            if (valueDecorated != null) {
+                return valueDecorated.length() >= 2 && valueDecorated.charAt(0) == '\''
+                        && valueDecorated.charAt(valueDecorated.length() - 1) == '\'';
+            }
+            return true;
+        }
+    }
+
+    /**
+     * Serialize {$link Arguments} into text representation. In case of empty arguments returns null.
+     */
+    public String serialize() {
+        StringBuilder sb = new StringBuilder();
+
+        for (Argument argument : getValues()) {
+            if (sb.length() > 0) {
+                sb.append(",");
+            }
+
+            sb.append(argument.getDigit());
+
+            if (argument.getValue() != null) {
+                sb.append("='");
+                sb.append(argument.getValue());
+                sb.append("'");
+            }
+        }
+
+        if (sb.toString().length() == 0) {
+            return null;
+        }
+
+        return sb.toString();
+    }
+
+    /**
+     * Creates a map that represents arguments.
+     */
+    public Map<Integer, String> toMap() {
+        Map<Integer, String> result = new HashMap<>();
+        for (Argument argument : getValues()) {
+            String value = argument.getValue();
+            result.put(argument.getDigit(), value);
+        }
+        return result;
+    }
+
+    /**
+     * Create a builder for {$link Arguments}.
+     */
+    public static ArgumentsBuilder builder() {
+        return new ArgumentsBuilder();
+    }
+
+    /**
+     * Builder for creating {$link Arguments}.
+     */
+    public static class ArgumentsBuilder {
+
+        private final List<Map<Integer, String>> keyValuePairs = new ArrayList<>();
```

Comment:
I'm not sure to get why using a List of Map here ?

Reply:
I tried to change just into Map but I realized that we drop unique-digit checking then. The checking (of uniqueness) is made in build method, not while adding arguments. So I had to preserve multiple key structure.

Alternatively I could use List of POJO Argument (local version of Argument, without validation in constructor) instead of list of map.